Weekly_Work/Week1/Code_changes-git/Metrics_Root_class.py

```diff
@@ -0,0 +1,202 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# In[1]:
+
+
+import json
+import datetime
+import matplotlib.pyplot as plt
+import pandas as pd
+
+
+# In[7]:
+
+
+class Metric():
+    
+    def __init__(self, path, since=None, until=None, source_code_exclude_list=None):
+
+        self.since = since
+        self.until = until
+        self.source_code_exclude_list = source_code_exclude_list
+        self.repo_urls = set()
+        
+        clean_commit_list = list()
+        clean_issue_list = list()
+        clean_pr_list = list()
+        
+        with open(path, 'r') as raw_data:
+            for line in raw_data:
+                line = json.loads(line)
+                self.repo_urls.add(line['origin'])
+                clean_line = dict()
+                if line['category'] == "commit":
+                    clean_line = self._clean_commit(line)
+
+                    if self._is_source_code(clean_line):
+                        clean_commit_list.append(clean_line)
+                    
+
+                elif line['category'] == "issue":
+                    if "pull_request" not in line['data']:
```

Comment:
This is the kind of code that would be run only for subclasses dealing with issues (or pull requests, below).

Reply:
The way I implemented it, it is possible for a Perceval data file to have commits, pull requests and hence, organize all data in  `clean_dict` and then every sub class can choose which category they want. 

Alternatively, in the root class, I could pass a `category` argument, which would then return only those data lines which satisfy that category (either commits, or pull requests or issues).

Let me know what you think.

Reply:
That's what I understood, but that forces you to include switch structures ("if then else") in the class, depending on functionality which would be, more cleanly, in children. I think the root class can (at least for now) just assume it reads "items" from a Perceval JSON file, with no assumption about the kind of item. Children that deal with commits, can do any processing (if needed), assuming their data structures include only commits.

Do you see any problem with this schema?

Reply:
If I understand correctly, the root class only reads every line in the JSON file and just adds them to a dataframe / list of dictionaries without any processing or cleaning. This is passed to the child class, which again goes through the dataframe/list choosing what it wants, then cleaning it and then computing the metric.

It's just my personal opinion, but since there are only three categories (issues, commits and pull requests), only a small number of if-else are used. 
I have no problem implementing what you suggest. 
Let me know if I understood it correctly (what I mentioned at the start of this comment.) :)

Reply:
Following from the comment above, if I understand correctly, a problem could arise here:
Since the root class just reads the file and adds rows to the dataframe, each type of data (either commit, pull request, or issue) will have different attributes (even for the raw data fetched by Perceval). It would be a messy affair to create a DataFrame with different data points, to be later cleaned by the child class. 
Of course, we could fix this by returning a list of dicts (for both the pandas and non-pandas immplementation) and this would later be cleaned by the child class.
Let me know what you think!

Reply:
@Polaris000,
This is what I think Jesus is proposing:
```
reference_implementations/ (parent directory)
   - RootDataFrame.py (Root class for dataframe object)
   - RootList.py (Root class for list of dict object)
   - MetricDataFrames/ (Metric implementations using dataframes)
      - Metric1.py
      - Metric2.py
   ...
   - MetricLists/ (Metric implementations for lists)
      - Metric1.py
      - Metric2.py
    ...
   - utils.py (general utility functions)
```

Now, each of the metrics are implemented using either the `git` datasource, or a `github_issues` data source or a `github_prs`  data source or the others that can be fetched using [Perceval](https://github.com/chaoss/grimoirelab-perceval#usage). So each of the `Metric` classes above will process the "items" file according to the datasource it can calculate the metric from.

The Root class will contain common functions which all the Metric classes will use. These functions can be inherited by the Metric classes from the Root class

Reply:
Yes, @aswanipranjal, that could be an structure. I would only make it flat for now, but we could also start exactly as you say.

WRT the different data sources and the root class, the root class can just have as an argument a dataframe, which would be produced by a function (or class) that just reads the JSON document and produces the dataframe (or list). i commented about this elsewhere: for producing the "right" dataframe from a list of Perceval dictionaries, each child class can implement a function (say "flatten()") that will extract the relevant data from the dictionary and produce a datafame row with it:

```
class Root():

   def __init__(self, list):
       ...
       for item in list:
           row = self.flatten(item)
           self.df.add(row)
       ....

class Metric_X()

    def flatten(self, item):
       '''Produce a flat row from the item (dictionary)'''
       .....
```

Reply:
> If I understand correctly, the root class only reads every line in the JSON file and just adds them to a dataframe / list of dictionaries without any processing or cleaning. This is passed to the child class, which again goes through the dataframe/list choosing what it wants, then cleaning it and then computing the metric.
> It's just my personal opinion, but since there are only three categories (issues, commits and pull requests), only a small number of if-else are used.

@Polaris000 since each metric will only use 1 data source, the if else will not be required while processing.

Reply:
Thanks @aswanipranjal!. I have already incorporated several changes and created #7 . I will make the required changes suggested by @jgbarah in #7 soon, and add a commit to the same pull request.

Reply:
@jgbarah @aswanipranjal, why is a flatten function required in the Root class? If I am passing a raw data list to it, and its child class will do the actual cleaning, all the root class is required to do is to convert the list of dictionaries to a dataframe, which is a simple `pandas.DataFrame(list_name)`

Reply:
@Polaris000, 
This is a sample from a git data source:
```
{
    "backend_name": "Git",
    "backend_version": "0.11.1",
    "category": "commit",
    "classified_fields_filtered": null,
    "data": {
        "Author": "Santiago Due\u00f1as <sduenas@bitergia.com>",
        "AuthorDate": "Fri May 17 17:18:57 2019 +0200",
        "Commit": "Santiago Due\u00f1as <sduenas@bitergia.com>",
        "CommitDate": "Fri May 17 17:18:57 2019 +0200",
        "Merge": "3360373 7891adb",
        "commit": "b6b6f6345c9e5e5a0877b8b40ffb8dd073f9ea3d",
        "files": [
            {
                "added": "30",
                "file": "perceval/backends/core/slack.py",
                "removed": "15"
            },
            {
                "added": "40",
                "file": "tests/test_slack.py",
                "removed": "11"
            }
        ],
        "message": "Merge branch 'handle-slack-timestamps-6-more' of 'https://github.com/valeriocos/perceval'\n\nMerges #530\nCloses #530",
        "parents": [
            "336037316f917c84c2785548336d2ee82ed5dc93",
            "7891adbc5d8c12bf13e3aad3652446cda93e6faa"
        ],
        "refs": [
            "HEAD -> refs/heads/master"
        ]
    },
    "origin": "https://github.com/grimoirelab/perceval.git",
    "perceval_version": "0.12.13",
    "tag": "https://github.com/grimoirelab/perceval.git",
    "timestamp": 1558266381.994731,
    "updated_on": 1558106337.0,
    "uuid": "1dddde0c2d265b5b9c4b3ac7f23641bd47018698"
}
```
Aas far as I can say, you won't be able to use it directly. You will have to _flatten_ this as in take the parts out of the `data` and get it to the same level as the other outer fields. So that you can create a data frame from the relevant fields

> The child classes which inherit from the Root class will implement this function and hence when the child class is initialized with the Perceval JSON files as input, the intialization step can clean the data according to the flatten function's implementation in the child class.
Look at this for a example of how this works

The Root class will not exactly convert the list of the JSON document above, the flatten implementation in the child class will clean it.

Reply:
I have a utils read method which reads the json file. And the root class object is instantiated using a datalist which it will flatten. I am planning to call the read method in each category class (Commit, Pull request, Issue), which will pass on the datalist to its parent, which is Metric Root class.
Thanks for the reply @aswaniprnjal! I didn't think the nesting would cause an issue, since I did the following:
```python
# Root class:
pandas.DataFrame(list_name)

# In Commit class:
raw_data_list = self.raw_df.to_dict(orient='records')
```
And noticed nothing strange. Will look into it!